Fix rubygem incompatibility errors

kozmotronik opened the pull request:
## Problem

- This PR is for resolving issue #55, a rubygems incompatibility problem that occurs at the installation phase.

## Solution

- The solution is to install a specific version of ruby gems instead of the latest one which is incompatible.

## PR Status

- [ ] Code finished and ready to be reviewed/tested
- [ ] The fix/enhancement were manually tested (if applicable)

## Automatic tests

Automatic tests can be triggered on https://ci-apps-dev.yunohost.org/ *after creating the PR*, by commenting "!testme", "!gogogadgetoci" or "By the power of systemd, I invoke The Great App CI to test this Pull Request!". (N.B. : for this to work you need to be a member of the Yunohost-Apps organization)

kozmotronik committed:
Fix rubygem incompatibility error at installation

kozmotronik commented:
I tested the installation on one my machine. It looks like this change breaks the systemd service? The service fails to start somehow but I couldn't see what the error is exactly. Here is a full log: https://paste.yunohost.org/raw/ameqidipoy.

Looking at `/var/log/redmine__2/redmine__2.log`:

```log
=> Booting WEBrick
=> Rails 7.2.2.1 application starting in production http://127.0.0.1:8097
=> Run `bin/rails server --help` for more startup options
[2025-12-10 12:20:53] INFO  WEBrick 1.9.2
[2025-12-10 12:20:53] INFO  ruby 3.1.7 (2025-03-26) [x86_64-linux]
[2025-12-10 12:20:53] INFO  Rackup::Handler::WEBrick::Server#start: pid=2419186 port=8097
[2025-12-10 12:25:50] FATAL SignalException: SIGTERM
	/opt/rbenv/versions/3.1.7/lib/ruby/gems/3.1.0/gems/webrick-1.9.2/lib/webrick/server.rb:173:in `select'
	/opt/rbenv/versions/3.1.7/lib/ruby/gems/3.1.0/gems/webrick-1.9.2/lib/webrick/server.rb:173:in `block in start'
	/opt/rbenv/versions/3.1.7/lib/ruby/gems/3.1.0/gems/webrick-1.9.2/lib/webrick/server.rb:32:in `start'
	/opt/rbenv/versions/3.1.7/lib/ruby/gems/3.1.0/gems/webrick-1.9.2/lib/webrick/server.rb:160:in `start'
	/opt/rbenv/versions/3.1.7/lib/ruby/gems/3.1.0/gems/rackup-2.3.1/lib/rackup/handler/webrick.rb:46:in `run'
	/opt/rbenv/versions/3.1.7/lib/ruby/gems/3.1.0/gems/rackup-2.3.1/lib/rackup/server.rb:341:in `start'
	/opt/rbenv/versions/3.1.7/lib/ruby/gems/3.1.0/gems/railties-7.2.2.1/lib/rails/commands/server/server_command.rb:38:in `start'
	/opt/rbenv/versions/3.1.7/lib/ruby/gems/3.1.0/gems/railties-7.2.2.1/lib/rails/commands/server/server_command.rb:145:in `block in perform'
	<internal:kernel>:90:in `tap'
	/opt/rbenv/versions/3.1.7/lib/ruby/gems/3.1.0/gems/railties-7.2.2.1/lib/rails/commands/server/server_command.rb:136:in `perform'
	/opt/rbenv/versions/3.1.7/lib/ruby/gems/3.1.0/gems/thor-1.4.0/lib/thor/command.rb:28:in `run'
	/opt/rbenv/versions/3.1.7/lib/ruby/gems/3.1.0/gems/thor-1.4.0/lib/thor/invocation.rb:127:in `invoke_command'
	/opt/rbenv/versions/3.1.7/lib/ruby/gems/3.1.0/gems/railties-7.2.2.1/lib/rails/command/base.rb:178:in `invoke_command'
	/opt/rbenv/versions/3.1.7/lib/ruby/gems/3.1.0/gems/thor-1.4.0/lib/thor.rb:538:in `dispatch'
	/opt/rbenv/versions/3.1.7/lib/ruby/gems/3.1.0/gems/railties-7.2.2.1/lib/rails/command/base.rb:73:in `perform'
	/opt/rbenv/versions/3.1.7/lib/ruby/gems/3.1.0/gems/railties-7.2.2.1/lib/rails/command.rb:71:in `block in invoke'
	/opt/rbenv/versions/3.1.7/lib/ruby/gems/3.1.0/gems/railties-7.2.2.1/lib/rails/command.rb:149:in `with_argv'
	/opt/rbenv/versions/3.1.7/lib/ruby/gems/3.1.0/gems/railties-7.2.2.1/lib/rails/command.rb:69:in `invoke'
	/opt/rbenv/versions/3.1.7/lib/ruby/gems/3.1.0/gems/railties-7.2.2.1/lib/rails/commands.rb:18:in `<top (required)>'
	bin/rails:4:in `require'
	bin/rails:4:in `<main>'
[2025-12-10 12:25:50] INFO  going to shutdown ...
[2025-12-10 12:25:50] INFO  Rackup::Handler::WEBrick::Server#start done.
Exiting
=> Booting WEBrick
=> Rails 7.2.2.2 application starting in production http://127.0.0.1:8097
=> Run `bin/rails server --help` for more startup options
/opt/rbenv/versions/3.4.7/lib/ruby/gems/3.4.0/gems/net-ldap-0.17.1/lib/net/ldap.rb:2: warning: ostruct was loaded from the standard library, but will no longer be part of the default gems starting from Ruby 3.5.0.
You can add ostruct to your Gemfile or gemspec to silence this warning.
[2025-12-13 08:50:11] INFO  WEBrick 1.9.2
[2025-12-13 08:50:11] INFO  ruby 3.4.7 (2025-10-08) [x86_64-linux]
[2025-12-13 08:50:11] INFO  Rackup::Handler::WEBrick::Server#start: pid=343931 port=8097
[2025-12-13 08:55:08] FATAL SignalException: SIGTERM
	/opt/rbenv/versions/3.4.7/lib/ruby/gems/3.4.0/gems/webrick-1.9.2/lib/webrick/server.rb:173:in 'IO.select'
	/opt/rbenv/versions/3.4.7/lib/ruby/gems/3.4.0/gems/webrick-1.9.2/lib/webrick/server.rb:173:in 'block in WEBrick::GenericServer#start'
	/opt/rbenv/versions/3.4.7/lib/ruby/gems/3.4.0/gems/webrick-1.9.2/lib/webrick/server.rb:32:in 'WEBrick::SimpleServer.start'
	/opt/rbenv/versions/3.4.7/lib/ruby/gems/3.4.0/gems/webrick-1.9.2/lib/webrick/server.rb:160:in 'WEBrick::GenericServer#start'
	/opt/rbenv/versions/3.4.7/lib/ruby/gems/3.4.0/gems/rackup-2.3.1/lib/rackup/handler/webrick.rb:46:in 'Rackup::Handler::WEBrick.run'
	/opt/rbenv/versions/3.4.7/lib/ruby/gems/3.4.0/gems/rackup-2.3.1/lib/rackup/server.rb:341:in 'Rackup::Server#start'
	/opt/rbenv/versions/3.4.7/lib/ruby/gems/3.4.0/gems/railties-7.2.2.2/lib/rails/commands/server/server_command.rb:38:in 'Rails::Server#start'
	/opt/rbenv/versions/3.4.7/lib/ruby/gems/3.4.0/gems/railties-7.2.2.2/lib/rails/commands/server/server_command.rb:145:in 'block in Rails::Command::ServerCommand#perform'
	<internal:kernel>:91:in 'Kernel#tap'
	/opt/rbenv/versions/3.4.7/lib/ruby/gems/3.4.0/gems/railties-7.2.2.2/lib/rails/commands/server/server_command.rb:136:in 'Rails::Command::ServerCommand#perform'
	/opt/rbenv/versions/3.4.7/lib/ruby/gems/3.4.0/gems/thor-1.4.0/lib/thor/command.rb:28:in 'Thor::Command#run'
	/opt/rbenv/versions/3.4.7/lib/ruby/gems/3.4.0/gems/thor-1.4.0/lib/thor/invocation.rb:127:in 'Thor::Invocation#invoke_command'
	/opt/rbenv/versions/3.4.7/lib/ruby/gems/3.4.0/gems/railties-7.2.2.2/lib/rails/command/base.rb:178:in 'Rails::Command::Base#invoke_command'
	/opt/rbenv/versions/3.4.7/lib/ruby/gems/3.4.0/gems/thor-1.4.0/lib/thor.rb:538:in 'Thor.dispatch'
	/opt/rbenv/versions/3.4.7/lib/ruby/gems/3.4.0/gems/railties-7.2.2.2/lib/rails/command/base.rb:73:in 'Rails::Command::Base.perform'
	/opt/rbenv/versions/3.4.7/lib/ruby/gems/3.4.0/gems/railties-7.2.2.2/lib/rails/command.rb:71:in 'block in Rails::Command.invoke'
	/opt/rbenv/versions/3.4.7/lib/ruby/gems/3.4.0/gems/railties-7.2.2.2/lib/rails/command.rb:149:in 'Rails::Command.with_argv'
	/opt/rbenv/versions/3.4.7/lib/ruby/gems/3.4.0/gems/railties-7.2.2.2/lib/rails/command.rb:69:in 'Rails::Command.invoke'
	/opt/rbenv/versions/3.4.7/lib/ruby/gems/3.4.0/gems/railties-7.2.2.2/lib/rails/commands.rb:18:in '<top (required)>'
	/opt/rbenv/versions/3.4.7/lib/ruby/3.4.0/bundled_gems.rb:82:in 'Kernel.require'
	/opt/rbenv/versions/3.4.7/lib/ruby/3.4.0/bundled_gems.rb:82:in 'block (2 levels) in Kernel#replace_require'
	bin/rails:4:in '<main>'
[2025-12-13 08:55:08] INFO  going to shutdown ...
[2025-12-13 08:55:08] INFO  Rackup::Handler::WEBrick::Server#start done.
Exiting

```

kozmotronik committed:
Comment out Ruby gem update and install commands

Comment out gem update and install commands since these commands are only required for Ruby 2.5 or earlier. See: https://www.redmine.org/projects/redmine/wiki/RedmineInstall#Step-4-Dependencies-installation

DeMiro5001 commented:
!testme

kozmotronik committed:
Set specific Ruby gem version in _common.sh

Set specific RubyGem version in order to comply the version requierement for the commonmarker plugin:

`Commonmarker 2.3.0 depends on RubytGems >= 3.4, < 4.A`

kozmotronik committed:
Include ostruct gem in Gemfile

Add ostruct gem to Gemfile to silence warning since it will no longer be part of the default gems starting from Ruby 4.0.0.

kozmotronik committed:
Update version to 6.1.1 and source URL in manifest

kozmotronik committed:
Update version badge in README.md to 6.1.1

kozmotronik committed:
Remove webrick gem from bundle installation

Removed webrick gem installation from bundle that causes "Could not find gem 'webrick (~> 1.7)' in locally installed gems" error.

kozmotronik committed:
Modify install script to append ostruct gem

Updated the installation script to append 'ostruct' gem to Gemfile.

kozmotronik committed:
Add extra gems to Gemfile.local

Added extra gems 'ostruct' and 'puma' to Gemfile.local for Redmine installation.

kozmotronik committed:
Change systemd service wait condition to 'Listening on http'

kozmotronik committed:
Remove 'puma' gem from Gemfile.local

Remove 'puma' gem from Gemfile.local since it already exists in the main Gemfile.

kozmotronik committed:
Add Puma gem and update systemd service timeout

* Add back Puma gem to Gemfile.local. I removed it since the installation was complaining about duplicate gems. But in the tests it couldn't find any server gem and complained as following:
WARNING - Could not find a server gem. Maybe you need to add one to the Gemfile?
WARNING -   gem "puma

* Also added 300 seconds timeout for systemd service start.

kozmotronik commented:
All issues are fixed step by step. You can see tha changes above in each commits.
The latest commit snapshot installed and tested in a virtualbox Yunohost. Installation was successful. See the full installation log here: https://paste.yunohost.org/raw/sezigahizo

Here is a screenshot of the installed instance:

<img width="940" height="757" alt="redmine-6 1 1-test-ok" src="https://github.com/user-attachments/assets/cc823fe8-53af-4347-babd-90cde1eaa104" />

So please, go ahead, trigger a CI test.

Thovi98 commented:
!testme

Thovi98 commented on scripts/install:
```diff
@@ -66,7 +77,7 @@ chown -R "$app:$app" "$install_dir"
 ynh_script_progression "Starting $app's systemd service..."
 
-ynh_systemctl --service="$app" --action="start" --wait_until="WEBrick::HTTPServer#start"
+ynh_systemctl --service="$app" --action="start" --wait_until="Listening on http" --timeout=300
```
This should be adapted in every script for coherence

kozmotronik replied:
You're right, let me check the other scripts.

Thovi98 commented on scripts/install:
```diff
@@ -21,6 +21,17 @@ ynh_config_add --template="database.example.yml" --destination="$install_dir/con
+# We should add extra gems to the Gemfile.local file according to the Redmine installation docs.
+cat <<EOF > "$install_dir/Gemfile.local"
+gem 'ostruct'
+gem 'puma'
```
192206 WARNING Your Gemfile lists the gem puma (>= 0) more than once.
192207 WARNING You should probably keep only one of them.
192207 WARNING Remove any duplicate entries and specify the gem only once.
192207 WARNING While it's not a problem now, it could cause errors if you change the version of one of them later.

Are you sure it's not already there?

Thovi98 replied:
And if it's needed, upgrade should at least be updated too

kozmotronik replied:
If you check the commit history, I added the puma to the Gemfile.local at first.

After getting those warnings I removed it and tested the installation. İnstallation failed at the service and server stage because it complained that puma cannot be find.

That's why I added it back. It actualy exist in the original Gemfile. But it is wrapped in a condition block which is for test environment only.

So what do you recommend? Maybe we better add webrick instead of puma?

Thovi98 replied:
It's fine for me like this :)

Thovi98 commented:
Thanks a lot! Great work!

kozmotronik committed:
Update the launching method of service

Update the launching method of service so that it be compatible with other scripts.

kozmotronik committed:
Update the launching method of service

Update the launching method of service so that it be compatible with other scripts.

kozmotronik committed:
Update the launching method of service

Update the launching method of service so that it be compatible with other scripts.

kozmotronik commented:
I've just updated the other three scripts that launch the redmine service. So it is supposed to pass the failed CI tests now.

> [!IMPORTANT]
> This is a breaking change for this YNH package since we switched from Webrick to Puma

Thovi98 commented:
Thanks, could you take care of the upgrade too?
Then I guess we’ll be fine.

kozmotronik commented:
> Thanks, could you take care of the upgrade too? Then I guess we’ll be fine.

I already did it in [this commit](https://github.com/kozmotronik/redmine_ynh/commit/270b30e4d97b662457258da05062ffc9c2ada156). Check it out please.

Thovi98 commented:
Yes, thank you!
But I talked about the update of the gem file:
https://ci-apps-dev.yunohost.org/ci/job/18164

kozmotronik committed:
Add extra gems to Gemfile.local

Add extra gems 'ostruct' and 'puma' to Gemfile.local for Redmine upgrade.

Thovi98 commented:
!testme

kozmotronik commented:
> Yes, thank you! But I talked about the update of the gem file: https://ci-apps-dev.yunohost.org/ci/job/18164

Oh, sorry, I am a noob Yunohost dev 😅 😅 

I've added the missing dependencies to the `upgrade` script. Please check it out here: https://github.com/kozmotronik/redmine_ynh/commit/acc0372334d112fcf44f05e7e4f6cda095219059

Thovi98 commented:
> Oh, sorry, I am a noob Yunohost dev 😅 😅

No worry, I learn every day also :D

Thovi98 committed:
remove failing test

Thovi98 commented:
Thanks for your great work!
